File: src/code/java/FindTheDifferenceOfTwoArrays.java
package code.java;

/**
 * https://leetcode.com/problems/find-the-difference-of-two-arrays/submissions/1624093269/?envType=study-plan-v2&envId=leetcode-75
 */

import java.util.*;

public class FindTheDifferenceOfTwoArrays {

    public static void main(String[] args) {
        int[] nums1 = {1,2,3};
        int[] nums2 = {2,4,6};

        List<List<Integer>> result = findDifference(nums1, nums2);
        System.out.println(result);
    }

    public static List<List<Integer>> findDifference(int[] nums1, int[] nums2) {
        Integer[] boxedArr1 = Arrays.stream(nums1).boxed().toArray(Integer[]::new);     // 배열 -> set 저장
        Set<Integer> set1 = new HashSet<>(Arrays.asList(boxedArr1));

        Integer[] boxedArr2 = Arrays.stream(nums2).boxed().toArray(Integer[]::new);     // 배열 -> set 저장
        Set<Integer> set2 = new HashSet<>(Arrays.asList(boxedArr2));

        Set<Integer> common = new HashSet<>();  // 중복되는 숫자 저장하는 set

        for(int s1 : set1){
            if(set2.contains(s1)){  // 중복되는 숫자 저장
                common.add(s1);
            }
        }

        List<List<Integer>> result = new ArrayList<>();
        result.add(set1.stream().filter(item -> !common.contains(item)).toList());  // set1에서 중복되는 숫자 제외하고 리스트 변환
        result.add(set2.stream().filter(item -> !common.contains(item)).toList());  // set2에서 중복되는 숫자 제외하고 리스트 변환

        return result;
    }
}

File: src/code/kotlin/FindTheDifferenceOfTwoArrays.kt
package code.kotlin

/**
 * https://leetcode.com/problems/find-the-difference-of-two-arrays/submissions/1624093269/?envType=study-plan-v2&envId=leetcode-75
 */

fun main() {
    val nums1 = intArrayOf(1, 2, 3)
    val nums2 = intArrayOf(2, 4, 6)

    val result = findDifference(nums1, nums2)
    println(result)
}

fun findDifference(nums1: IntArray, nums2: IntArray): List<List<Int>> {
    val set1 = nums1.toSet()    // 배열 -> set 저장
    val set2 = nums2.toSet()    // 배열 -> set 저장

    val common = set1.intersect(set2)   // 중복되는 숫자 저장하는 set

    val uniqueToNums1 = set1.filterNot { it in common }     // set1에서 중복되는 숫자 제외하고 리스트 변환
    val uniqueToNums2 = set2.filterNot { it in common }     // set2에서 중복되는 숫자 제외하고 리스트 변환

    return listOf(uniqueToNums1, uniqueToNums2)
}